optuna/integration/pytorch_distributed.py

```diff
@@ -0,0 +1,246 @@
+                )
+
+            assert trial is None, "error message"
+        self._delegate = trial
+        self._device = device
+
+    def suggest_float(
+        self,
+        name: str,
+        low: float,
+        high: float,
+        *,
+        step: Optional[float] = None,
+        log: bool = False,
+    ) -> float:
+        def func() -> float:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_float(name, low, high, step=step, log=log)
+
+        return self._call_and_communicate(func, torch.float)
+
+    def suggest_uniform(self, name: str, low: float, high: float) -> float:
+        def func() -> float:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_uniform(name, low, high)
+
+        return self._call_and_communicate(func, torch.float)
+
+    def suggest_loguniform(self, name: str, low: float, high: float) -> float:
+        def func() -> float:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_loguniform(name, low, high)
+
+        return self._call_and_communicate(func, torch.float)
+
+    def suggest_discrete_uniform(self, name: str, low: float, high: float, q: float) -> float:
+        def func() -> float:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_discrete_uniform(name, low, high, q=q)
+
+        return self._call_and_communicate(func, torch.float)
+
+    def suggest_int(self, name: str, low: int, high: int, step: int = 1, log: bool = False) -> int:
+        def func() -> float:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_int(name, low, high, step=step, log=log)
+
+        return self._call_and_communicate(func, torch.int)
+
+    def suggest_categorical(self, name: str, choices: Sequence["CategoricalChoiceType"]) -> Any:
+        def func() -> CategoricalChoiceType:
+
+            assert self._delegate is not None
+            return self._delegate.suggest_categorical(name, choices)
+
+        return self._call_and_communicate_obj(func)
+
+    def report(self, value: float, step: int) -> None:
+
+        if dist.get_rank() == 0:
+            assert self._delegate is not None
+            self._delegate.report(value, step)
+        dist.barrier()
+
+    def should_prune(self) -> bool:
+        def func() -> bool:
+
+            assert self._delegate is not None
+            # Some pruners return numpy.bool_, which is incompatible with bool.
+            return bool(self._delegate.should_prune())
+
+        # torch.bool seems to be the correct type, but the communication fails
+        # due to the RuntimeError.
+        return self._call_and_communicate(func, torch.uint8)
+
+    def set_user_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self._delegate is not None
+            self._delegate.set_user_attr(key, value)
+        dist.barrier()
```

Comment:
When only partial nodes call this method (e.g., calling from only the rank 0 node), execution stops silently. I recommend mentioning this behavior in the document. I think it is useful to add a `barrier` argument to control whether to synchronize.

Reply:
Thank you for your suggestion. I added a note to call the methods from all workers in the commit https://github.com/optuna/optuna/pull/2303/commits/eea96c85f8f4321d64538cda32f028278a94c67d.

> I think it is useful to add a barrier argument to control whether to synchronize.

When I used `ChainerMNTrial`, I sometimes got errors without `barrier` due to the timing issue. So, I'm a bit conservative to add an option to switch `barrier`. I'd like to create a new PR if I work on it.

Reply:
I agree to use `barrier` as default behavior, but users may want to write code like below.
```python
if dist.get_rank() == 0:
    trial.set_user_attr(..., barrier=False)
dist.barrier()
```
In this case, the synchronization is done by users. I think it is reasonable to discuss this feature in another issue or PR.

Reply:
Similar to the above comment, if an exception is raised in `self._delegate.set_user_attr`, execution silently stops. It is better to catch the exception and broadcast it to all nodes.

Reply:
Thanks. I added the code to broadcast exceptions to all nodes and unit tests in the commit https://github.com/optuna/optuna/pull/2303/commits/39070bef147e03ebc9f8c6e04a8662ac6c223309. PTAL.

Reply:
LGTM!